src/qibo/parameter.py

```diff
@@ -0,0 +1,134 @@
+                features=[7.0],
+                trainable=[1.5, 2.0, 3.0],
+            )
+
+        partial_derivative = param.get_partial_derivative(3)
+
+        param.update_parameters(trainable=[15.0, 10.0, 7.0], feature=[5.0])
+        param_value = param()
+
+
+    Args:
+        func (function): lambda function which builds the gate parameter. If both features and trainable parameters
+        compose the function, it must be passed by first providing the features and then the parameters, as
+        described in the code example above.
+        features (list or np.ndarray): array containing possible input features x.
+        trainable (list or np.ndarray): array with initial trainable parameters theta.
+        nofeatures (bool): flag to explicitly ban the updating of the features. This simplifies the task of updating Parameter objects simultaneously when some have embedded features and some do not.
+    """
+
+    def __init__(self, func, trainable=[], features=[]):
+        self.trainable = trainable
+        self.features = features
+
+        if self.nfeat + self.nparams != func.__code__.co_argcount:
+            raise_error(
+                TypeError,
+                f"{self.nfeat + self.nparams} parameters are provided, but {func.__code__.co_argcount} are required, please initialize features and trainable according to the defined function.",
+            )
+        # lambda function
+        self.lambdaf = func
+
+        # calculate derivatives
+        # maybe here use JAX ?
+        self.derivatives = calculate_derivatives(func=self.lambdaf)
+
+    def __call__(self, features=None, trainable=None):
+        """Return parameter value with given features and/or trainable."""
+
+        params = []
+
+        if features is None:
+            params.extend(self.features)
+        else:
+            if len(features) != self.nfeat:
+                raise_error(
+                    TypeError,
+                    f"The number of features provided is not compatible with the problem's dimensionality, which is {self.nfeat}.",
+                )
+            else:
+                params.extend(features)
+        if trainable is None:
+            params.extend(self.trainable)
+        else:
+            if len(trainable) != self.nparams:
+                raise_error(
+                    TypeError,
+                    f"The number of trainable provided is different from the number of required parameters, which is {self.nparams}.",
+                )
+            else:
+                params.extend(trainable)
+
+        return self.lambdaf(*params)
+
+    @property
+    def nparams(self):
+        """Returns the number of trainable parameters"""
+        try:
+            return len(self.trainable)
+        except TypeError:
+            return 0
+
+    @property
+    def nfeat(self):
+        """Returns the number of features"""
+        return len(self.features) if isinstance(self.features, list) else 0
```

Comment:
In these cases, `try... except...` is usually the way to go (more robust).
```suggestion
        try:
            return len(self.features)
        except TypeError:
            return 0
```